File: frontend/src/layout/preferences.ts
import type {
  CardId,
  LayoutName,
  ModuleId,
  Preferences,
  PreferencesPatch,
  SectionId,
  User,
} from "../api/types";

/**
 * The account's layout preferences on the client (Epic 33, AD-49).
 *
 * The server always answers resolved, so the only default the client needs is for a
 * server older than migration 0025, which sends no `preferences` at all. It is the app as
 * it was, and it must stay equal to the server's `resolve({})` — both sides pin the same
 * literal in their tests.
 */
const SECTIONS: [SectionId, "bar" | "top"][] = [
  ["dashboard", "bar"],
  ["entries", "bar"],
  ["habits", "bar"],
  ["stock", "bar"],
  ["gym", "bar"],
  ["plan", "top"],
  ["grow", "top"],
  ["recipes", "top"],
];
export const MODULES: ModuleId[] = ["habits", "books", "mood", "stock", "gym", "recipes", "notes"];
export const CARDS: CardId[] = [
  "stats",
  "pending",
  "reading",
  "quote",
  "restock",
  "budgets",
  "savings",
  "trends",
  "categories",
];
export const LAYOUTS: LayoutName[] = ["phone", "desktop"];

function defaultLayout() {
  return {
    tabs: SECTIONS.map(([id, slot]) => ({ id, slot })),
    cards: CARDS.map((id) => ({ id, on: true })),
  };
}

export const DEFAULT_PREFERENCES: Preferences = {
  modules: Object.fromEntries(MODULES.map((id) => [id, true])) as Record<ModuleId, boolean>,
  phone: defaultLayout(),
  desktop: defaultLayout(),
};

/** The account's preferences, or the app as it was when the server predates them. */
export function preferencesOf(user: User | null | undefined): Preferences {
  return user?.preferences ?? DEFAULT_PREFERENCES;
}

/** What the server does with a patch: each top-level key present replaces that subtree. */
export function applyPatch(prefs: Preferences, patch: PreferencesPatch): Preferences {
  return { ...prefs, ...patch };
}

/**
 * Saves preferences one request at a time, and shows the person their latest choice at once.
 *
 * **One request in flight, never two.** The server applies patches in the order they
 * *arrive*, not the order they were asked; two in flight could land the older one last and
 * leave the account disagreeing with the screen. So a change made while a request is out is
 * queued, and changes queued together are merged — they replace whole subtrees, so the
 * merge is exactly what sending them in turn would have done — and sent as one.
 *
 * **What is shown** is always the last state the server confirmed with the in-flight and
 * queued patches laid over it. A request that fails drops its own patch and nothing else:
 * the screen falls back to the confirmed state plus whatever is still queued, and the
 * promise each caller got for that patch rejects, so the control that asked can say so.
 */
export class PreferenceSaver {
  private confirmed: Preferences;
  private inflight: PreferencesPatch | null = null;
  private queued: PreferencesPatch | null = null;
  private waiting: { resolve: () => void; reject: (error: unknown) => void }[] = [];

  constructor(
    initial: Preferences,
    private readonly send: (patch: PreferencesPatch) => Promise<Preferences>,
    private readonly onChange: (shown: Preferences) => void,
  ) {
    this.confirmed = initial;
  }

  /** What the screen should show now. */
  get shown(): Preferences {
    return applyPatch(applyPatch(this.confirmed, this.inflight ?? {}), this.queued ?? {});
  }

  /** The server said so from elsewhere (a profile refresh): adopt it under anything pending. */
  confirm(prefs: Preferences): void {
    this.confirmed = prefs;
    this.onChange(this.shown);
  }

  update(patch: PreferencesPatch): Promise<void> {
    this.queued = { ...this.queued, ...patch };
    const done = new Promise<void>((resolve, reject) => {
      this.waiting.push({ resolve, reject });
    });
    this.onChange(this.shown);
    if (this.inflight === null) void this.drain();
    return done;
  }

  private async drain(): Promise<void> {
    while (this.queued !== null) {
      const patch = this.queued;
      const callers = this.waiting;
      this.queued = null;
      this.waiting = [];
      this.inflight = patch;
      try {
        this.confirmed = await this.send(patch);
        this.inflight = null;
        this.onChange(this.shown);
        for (const caller of callers) caller.resolve();
      } catch (error) {
        this.inflight = null;
        this.onChange(this.shown);
        for (const caller of callers) caller.reject(error);
      }
    }
  }
}

File: backend/app/services/preferences.py
"""The account's layout preferences (Epic 33, AD-49).

One ``jsonb`` column on ``users``, **sparse**: a key that is not stored means "the
default". So a card or a section added by a later epic appears for everyone without a data
migration, and the one function that fills the gaps is :func:`resolve`.

What is stored is what the client sent, validated here; what is returned is always
resolved, so the client never has to guess a default. Shape checks (types, lengths, the
two slot names) are the request schema's; everything that needs to know the catalogue
below — unknown ids, duplicates, completeness, the phone's caps — is here, with a code.
"""

import uuid

from sqlalchemy import bindparam, update
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import Session

from app.core.errors import Invalid
from app.models.user import User

#: The eight sections, in their default order, with the slot each starts in. The first
#: five are the bottom tab bar on a phone (``.nav`` on a desktop); the last three are the
#: top bar (``.nav-extra``). Reproduces the app as it was before Epic 33.
SECTIONS: tuple[tuple[str, str], ...] = (
    ("dashboard", "bar"),
    ("entries", "bar"),
    ("habits", "bar"),
    ("stock", "bar"),
    ("gym", "bar"),
    ("plan", "top"),
    ("grow", "top"),
    ("recipes", "top"),
)

#: The budget itself. These can be moved, never switched off.
CORE_SECTIONS = frozenset({"dashboard", "entries", "plan", "grow"})

#: What can be switched off. Off hides the UI only; data and endpoints are untouched.
MODULES: tuple[str, ...] = ("habits", "books", "mood", "stock", "gym", "recipes", "notes")

#: Dashboard cards in today's render order, all shown by default.
CARDS: tuple[str, ...] = (
    "stats",
    "pending",
    "reading",
    "quote",
    "restock",
    "budgets",
    "savings",
    "trends",
    "categories",
)

LAYOUTS: tuple[str, ...] = ("phone", "desktop")

#: The phone top bar's content box is 335px at a 375px viewport, and French is the wide
#: language (Epic 27). Five tabs and three top links are what was measured to fit.
PHONE_CAPS = {"bar": 5, "top": 3}

_SECTION_IDS = tuple(section_id for section_id, _ in SECTIONS)
_DEFAULT_SLOT = dict(SECTIONS)


def _merge(stored: list[dict], catalogue: tuple[str, ...], make) -> list[dict]:
    """Stored order, minus ids that no longer exist, plus ids it lacks at their default place.

    A missing id is inserted after its nearest default predecessor that is already in the
    list, or first if it has none — so a card added after "reading" in a later epic lands
    after "reading" wherever the person moved it, instead of at the bottom.
    """
    known = set(catalogue)
    result: list[dict] = []
    seen: set[str] = set()
    for item in stored:
        item_id = item.get("id") if isinstance(item, dict) else None
        if item_id in known and item_id not in seen:
            result.append(item)
            seen.add(item_id)

    for index, item_id in enumerate(catalogue):
        if item_id in seen:
            continue
        position = 0
        for earlier in reversed(catalogue[:index]):
            if earlier in seen:
                position = next(i for i, it in enumerate(result) if it["id"] == earlier) + 1
                break
        result.insert(position, make(item_id))
        seen.add(item_id)
    return result


def _resolve_layout(stored: object) -> dict:
    layout = stored if isinstance(stored, dict) else {}
    tabs = _merge(
        [
            {"id": tab["id"], "slot": tab["slot"]}
            for tab in layout.get("tabs") or []
            if isinstance(tab, dict) and isinstance(tab.get("id"), str)
            and tab.get("slot") in ("bar", "top")
        ],
        _SECTION_IDS,
        lambda section_id: {"id": section_id, "slot": _DEFAULT_SLOT[section_id]},
    )
    cards = _merge(
        [
            {"id": card["id"], "on": card["on"]}
            for card in layout.get("cards") or []
            if isinstance(card, dict) and isinstance(card.get("id"), str)
            and isinstance(card.get("on"), bool)
        ],
        CARDS,
        lambda card_id: {"id": card_id, "on": True},
    )
    return {"tabs": tabs, "cards": cards}


def resolve(stored: object) -> dict:
    """The full preferences for a stored value, defaults filled in. Never raises: a stored
    value from an older catalogue is read, not refused."""
    prefs = stored if isinstance(stored, dict) else {}
    modules = prefs.get("modules") if isinstance(prefs.get("modules"), dict) else {}
    return {
        "modules": {
            module: modules[module] if isinstance(modules.get(module), bool) else True
            for module in MODULES
        },
        **{layout: _resolve_layout(prefs.get(layout)) for layout in LAYOUTS},
    }


def _no_duplicates(ids: list[str], what: str) -> None:
    seen: set[str] = set()
    for item_id in ids:
        if item_id in seen:
            raise Invalid(f"{what} {item_id!r} is listed twice", "pref_duplicate")
        seen.add(item_id)


def _check_modules(modules: dict[str, bool]) -> None:
    for module in modules:
        if module in CORE_SECTIONS:
            raise Invalid(f"{module!r} is part of the budget and cannot be turned off",
                          "pref_core_module")
        if module not in MODULES:
            raise Invalid(f"no module called {module!r}", "pref_unknown_id")


def _check_layout(name: str, layout: dict) -> None:
    tabs = layout.get("tabs")
    if tabs is not None:
        ids = [tab["id"] for tab in tabs]
        for section_id in ids:
            if section_id not in _DEFAULT_SLOT:
                raise Invalid(f"no section called {section_id!r}", "pref_unknown_id")
        _no_duplicates(ids, "section")
        # Complete, unlike the cards: a tab list that dropped a section would leave it
        # nowhere to be reached, and "hide" is what modules are for.
        if len(ids) != len(_SECTION_IDS):
            raise Invalid("the tab list must name every section once", "pref_incomplete")
        if name == "phone":
            for slot, cap in PHONE_CAPS.items():
                if sum(tab["slot"] == slot for tab in tabs) > cap:
                    raise Invalid(f"a phone fits at most {cap} in the {slot}", "pref_slot_full")

    cards = layout.get("cards")
    if cards is not None:
        ids = [card["id"] for card in cards]
        for card_id in ids:
            if card_id not in CARDS:
                raise Invalid(f"no dashboard card called {card_id!r}", "pref_unknown_id")
        _no_duplicates(ids, "card")


def validate(patch: dict) -> None:
    """Refuse, with a code, anything the catalogue does not allow. ``patch`` holds only
    the top-level keys the request sent, already shape-checked by the schema."""
    if "modules" in patch:
        _check_modules(patch["modules"])
    for layout in LAYOUTS:
        if layout in patch:
            _check_layout(layout, patch[layout])


def update_preferences(session: Session, user_id: uuid.UUID, patch: dict) -> None:
    """Replace each top-level key the patch carries; leave the others as stored.

    One ``UPDATE ... SET preferences = preferences || :patch``, not a read-modify-write:
    two tabs saving ``modules`` and ``phone`` at the same moment both land. Refusals are
    raised before anything is written.
    """
    validate(patch)
    if not patch:
        return
    session.execute(
        update(User)
        .where(User.id == user_id)
        .values(preferences=User.preferences.op("||")(bindparam("patch", patch, type_=JSONB)))
    )
    session.flush()
